src/python/pants/backend/rust/lint/rustfmt/rules_integration_test.py

```diff
@@ -0,0 +1,177 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from textwrap import dedent
+
+import pytest
+
+from pants.backend.rust.lint.rustfmt.rules import RustfmtFieldSet, RustfmtRequest
+from pants.backend.rust.lint.rustfmt.rules import rules as rustfmt_rules
+from pants.backend.rust.target_types import RustCrateTarget
+from pants.backend.rust.util_rules import toolchains
+from pants.core.goals.fmt import FmtResult
+from pants.core.util_rules import source_files, system_binaries
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.addresses import Address
+from pants.engine.fs import CreateDigest, Digest, FileContent
+from pants.engine.internals.native_engine import Snapshot
+from pants.engine.target import Target
+from pants.testutil.rule_runner import QueryRule, RuleRunner
+
+
+@pytest.fixture()
+def rule_runner() -> RuleRunner:
+    rule_runner = RuleRunner(
+        target_types=[RustCrateTarget],
+        rules=[
+            *rustfmt_rules(),
+            *toolchains.rules(),
+            *source_files.rules(),
+            *system_binaries.rules(),
+            QueryRule(FmtResult, (RustfmtRequest.Batch,)),
+            QueryRule(SourceFiles, (SourceFilesRequest,)),
+        ],
+    )
+    rule_runner.set_options([], env_inherit={"PATH"})
+    return rule_runner
+
+
+GOOD_FILE = dedent(
+    """\
+    use Foo::Bar;
+    fn grok(s: &str) {}
+    """
+)
+
+BAD_FILE = dedent(
+    """\
+    use  Foo::Bar;
+    fn grok(s: &str) {
+    println!("Hello World {}!", s);
+    }
+    """
+)
+
+FIXED_BAD_FILE = dedent(
+    """\
+    use Foo::Bar;
+    fn grok(s: &str) {
+        println!("Hello World {}!", s);
+    }
+    """
+)
+
+
+def run_rustfmt(
+    rule_runner: RuleRunner,
+    targets: list[Target],
+    *,
+    extra_args: list[str] | None = None,
+) -> FmtResult:
+    rule_runner.set_options(extra_args or (), env_inherit={"PATH"})
+    field_sets = [RustfmtFieldSet.create(tgt) for tgt in targets]
+    input_sources = rule_runner.request(
+        SourceFiles,
+        [
+            SourceFilesRequest(field_set.sources for field_set in field_sets),
+        ],
+    )
+    fmt_result = rule_runner.request(
+        FmtResult,
+        [
+            RustfmtRequest.Batch(
+                "",
+                input_sources.snapshot.files,
+                partition_metadata=None,
+                snapshot=input_sources.snapshot
+            ),
+        ],
+    )
+    return fmt_result
+
+
+def get_snapshot(rule_runner: RuleRunner, source_files: dict[str, str]) -> Snapshot:
```

Comment:
This method can be deleted. There was a refactor to switch tests to use `rule_runner.make_snapshot`.